Set for Android and Library alignment

YoliFD opened the pull request:
# Summary

I have found there are a few issues with the library which was preventing it from fully fitting my use case. Some of these issues are now also raised by others. My main concern was making the library more consistent and generic.

With these changes, I have addressed the following

1.  Moved away from using `ForwardingCookieHandler`  https://github.com/react-native-community/cookies/issues/19

The original library was tied to an RN class which actually just abstracted interactions with the standard libraries for handling cookies on Android. There is no real need for that and actually implementing the functionality directly is cleaner (the aforementioned class was making transformations around cookie setting for setFromResponse and under the hood using what we actually need for a normal set either way)

2. Adding support for `.set()` on Android https://github.com/react-native-community/cookies/issues/25

This meant that to align the protocols I had to introduce a URL parameter for both platforms but you can now set cookies on Android as well.

3. Aligned *return types* for both native implementations https://github.com/react-native-community/cookies/issues/11

I use TypeScript day to day and it was a bit of an inconvenience when the promises returned different types for Android and iOS. I have now aligned the return types, made sure the promise rejections are clearly called out and illustrated the change in the types index. 

4. Refactored `get()` to return a map of cookie objects (similar to this https://github.com/react-native-community/cookies/issues/29 feature request)

I also took this opportunity to allow get to return a cookie object fulfilled by the native libraries, instead of just the value itself but I have kept the return type to a map of name to cookie object. This can be easily changed but it does allow quick access to a cookie by name without iterating a list.

5. Looked into implementation of `.getAll` and `.clearByName` for Android (similar to https://github.com/react-native-community/cookies/issues/26)

Sadly both of these functions are not something that is natively available for Android. I explored all sorts of options here but there isn't much we can do that will work reliably. 

It is however important note that `get()` for Android will actually give a list of all cookies with domain related to the host of the url so it effectively acts as a getAll but for a single url. The Android system behaviour buckets the cookies based on a url (similarly to what you will see when you inspect cookies on Chrome) but does not allow for crawling through absolutely everything.

6. Introduced 2 new properties to a cookie `secure`, `httpOnly` 

Introduced support for secure and http only cookies in set too. This was easy enough as both native implementations support these 2 properties. 

7. Looked into adding SameSite support https://github.com/react-native-community/cookies/issues/30

While looking to extend the cookie object, I tried to find a way to add same site support but the native platform support was a bit mismatched on this one so I left it out for now. 

----

As it may have not been evident above, in this endeavour I have introduced *some braking changes* so I am more than happy to discuss and alter what is necessary to get these changes in and help others.

## Test Plan

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI. -->

### What's required for testing (prerequisites)?

iOS and Android simulators

### What are the steps to reproduce (after prerequisites)?

`README` and Typescript index files have been changed to reflect. Use the `README` instructions to check the interactions with all library methods as this is I realise quite a big change. 

## Compatibility

| OS      | Implemented |
| ------- | :---------: |
| iOS     |    ✅     |
| Android |    ✅     |

## Checklist

<!-- Check completed item, when applicable, via: [X] -->

- [x] I have tested this on a device and a simulator
- [x] I added the documentation in `README.md`
- [ ] I mentioned this change in `CHANGELOG.md`
- [x] I updated the typed files (TS and Flow)
- [ ] I added a sample use of the API in the example project (`example/App.js`)

Aure77 commented on README.md:
```diff
@@ -84,119 +86,143 @@ protected List<ReactPackage> getPackages() {
+    console.log('CookieManager.clearAll =>', success);
   });
 
 // clear a specific cookie by its name (IOS ONLY)
```
still iOS only ?

Aure77 commented on android/build.gradle:
```diff
@@ -73,10 +76,12 @@ repositories {
 dependencies {
     //noinspection GradleDynamicVersion
     implementation 'com.facebook.react:react-native:+'  // From node_modules
+    implementation("com.squareup.okhttp3:okhttp:4.4.0") // necessary for the cookie setting to work
```
React Native core use okhttp 3.12.1, there is no collision to use this version (4.4.0) ?

safaiyeh replied:
From the team

> ...might cause breakage if API changed
> I don’t expect FB to upgrade Okhttp because it dropped Android 4.x support

We should find an alternative.

YoliFD replied:
Fair point guys. In such case we have 2 options: 

A. I can match the same version as RN uses. That shouldn't be a problem. 

B. Find a different option for the cookie string make up. 

I'll start with A and look at options for B.

YoliFD replied:
Got rid of okhttp as requested. I tested more thoroughly and indeed it would cause issues.

Aure77 commented on android/src/main/java/com/reactnativecommunity/cookies/CookieManagerModule.java:
```diff
@@ -63,36 +78,180 @@ public void getFromResponse(String url, Promise promise) throws URISyntaxExcepti
+                cookieMap.putBoolean("httpOnly", cookie.httpOnly());
+
+                long expires = cookie.expiresAt();
+                cookieMap.putString("expiration", new Date(expires).toString());
```
expiration date should be rfc1123 compliant.
Since `toGMTString()` is deprecated, you should format with this pattern : `EEE, dd MMM yyyy HH:mm:ss zzz`
use `DateTimeFormatter.RFC_1123_DATE_TIME.format(...)` if Java 8 is the min supported java in this project. 
Or use SimpleDateFormatter otherwise :
```java
DateFormat df = new SimpleDateFormat("EEE, dd MMM yyyy HH:mm:ss zzz", java.util.Locale.US);
df.setTimeZone(TimeZone.getTimeZone("GMT"));
cookieMap.putString("expiration", df.format(expires));
```

YoliFD replied:
Will get that amended and testes as soon as I can.

YoliFD replied:
I have addressed this in the format of the cookie string for setting. However for the cookie object passed in for set and the cookie objects returned from get I have matched the iOS formatting `yyyy-MM-dd'T'HH:mm:ss.SSSZZZZZ`

YoliFD replied:
Since we are making breaking changes, there is an argument to using the same format across the board and complying with the cookie string date format i.e. RFC 1123. 
What do you guys think @Aure77 and @safaiyeh ?

safaiyeh replied:
Yeah let's keep it consistent across the board

YoliFD replied:
Cool, will be able to dedicate some time tomorrow and will swap them all out to use the RFC 1123 format. Currently we are not rejecting cookies if the date fails, which is probably not quite right. I will add that in too. Shall shout out when that is sorted :)

safaiyeh commented:
This is awesome! Thanks for all your hard work @YoliFD, I'll dedicate sometime today to review this.

safaiyeh commented on ios/RNCookieManagerIOS/RNCookieManagerIOS.m:
```diff
@@ -42,64 +42,45 @@ + (BOOL)requiresMainQueueSetup
     resolver:(RCTPromiseResolveBlock)resolve
     rejecter:(RCTPromiseRejectBlock)reject)
 {
-    NSString *name = [RCTConvert NSString:props[@"name"]];
-    NSString *value = [RCTConvert NSString:props[@"value"]];
-    NSString *path = [RCTConvert NSString:props[@"path"]];
-    NSString *domain = [RCTConvert NSString:props[@"domain"]];
-    NSString *origin = [RCTConvert NSString:props[@"origin"]];
-    NSString *version = [RCTConvert NSString:props[@"version"]];
-    NSDate *expiration = [RCTConvert NSDate:props[@"expiration"]];
-
-    NSMutableDictionary *cookieProperties = [NSMutableDictionary dictionary];
-    [cookieProperties setObject:name forKey:NSHTTPCookieName];
-    [cookieProperties setObject:value forKey:NSHTTPCookieValue];
-    [cookieProperties setObject:path forKey:NSHTTPCookiePath];
-    if (!isEmpty(domain)) {
-        [cookieProperties setObject:domain forKey:NSHTTPCookieDomain];
-    }
-    if (!isEmpty(origin)) {
-        [cookieProperties setObject:origin forKey:NSHTTPCookieOriginURL];
-    }
-    if (!isEmpty(version)) {
-         [cookieProperties setObject:version forKey:NSHTTPCookieVersion];
-    }
-    if (!isEmpty(expiration)) {
-         [cookieProperties setObject:expiration forKey:NSHTTPCookieExpires];
-    }
-
-    NSHTTPCookie *cookie = [NSHTTPCookie cookieWithProperties:cookieProperties];
```
Nice clean up

safaiyeh commented:
Hi @YoliFD I left some comments!

The domain field may no longer be necessary. Since the URL host has validation, it creates weird cases like URL host provided + domain provided.. one is valid, one invalid.. etc.

What do you think about making URL host required and removing the domain from the cookie object?

YoliFD commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@react-native-community/cookies",
-  "version": "2.0.8",
+  "version": "2.1.0",
```
Certainly! Sorry I thought I had already done that.

YoliFD commented on android/src/main/java/com/reactnativecommunity/cookies/CookieManagerModule.java:
```diff
@@ -63,36 +78,180 @@ public void getFromResponse(String url, Promise promise) throws URISyntaxExcepti
+                cookieMap.putString("value", cookie.value());
+                cookieMap.putString("domain", cookie.domain());
+                cookieMap.putString("path", cookie.path());
+                // no version unavailable from this interface
+                // cookieMap.putInt("version", cookie.getVersion());
```
Will do.

YoliFD commented on android/src/main/java/com/reactnativecommunity/cookies/CookieManagerModule.java:
```diff
@@ -63,36 +78,180 @@ public void getFromResponse(String url, Promise promise) throws URISyntaxExcepti
+                cookieMap.putBoolean("httpOnly", cookie.httpOnly());
+
+                long expires = cookie.expiresAt();
+                cookieMap.putString("expiration", new Date(expires).toString());
```
Sure thing

YoliFD commented on android/src/main/java/com/reactnativecommunity/cookies/CookieManagerModule.java:
```diff
@@ -63,36 +78,180 @@ public void getFromResponse(String url, Promise promise) throws URISyntaxExcepti
+        } else if (extractedDomain != null && !extractedDomain.isEmpty()) {
+            cookieBuilder.domain(extractedDomain);
+        } else {
+            // assume something went terribly wrong here and no cookie can be created
+            return null;
```
The null return from here should surface all the way up in the set method where at line 57 there is a check against it and a promise rejection due to invalid values. I opted for a promise rejection as it seemed the most appropriate way to notify. Do we want to add something else in addition to that ?

YoliFD replied:
I have included further messaging so it is more clear what the problems are.

YoliFD commented:
@safaiyeh and @Aure77  Thank you both for the comments 🙇 . I will look at making the changes as soon as I can.

In response to the question: 

> What do you think about making URL host required and removing the domain from the cookie object?

I considered situations where we might have mismatch between the host in the url (which can be just the full url, not just the host itself) and the domain. Which led me to the same thought about removing the domain but then considered situation where: 

We have url `http://subA.subB.foo.com` for which will by default the cookie will be attributed to `subA.subB.foo.com` (please correct me if I am wrong) **BUT**  a dev might actually want to attribute it to  `.subB.foo.com` or even any subdomains under the main `.foo.com`. 

I think you have a point about making sure that the domain and host match up when making the cookie strings on native so I will look at adding more validation around that (and appropriate messages for the developers.

safaiyeh commented:
@YoliFD Right, I'm trying to think of the situation where `http://subA.subB.foo.com` is the URL host and I want to attribute the cookie to subB, would just requiring to use `http://subB.foo.com` would be cleaner? I haven't needed to use specific domain cookies yet so maybe I am missing a popular use case.

safaiyeh commented:
Hi @YoliFD! Any updates on this? If you want, I can take it up from here!

YoliFD commented:
Hi @safaiyeh  Apologies for the delay. I have been struggling for time lately but have pushed a few changes. Will work to finish the changes requested tonight and give you a shout.

YoliFD commented:
I have now addressed all the comments and completely got rid of OKHTTP by matching the functionality necessary. 

I have left the domain field for cookie as well as the url but have ensure proper messaging is in place when these are mismatched and the cookie set will be rejected. 

We have a pending question on date formatting, if we want to make it the same across the board. 

I'll dedicate some more time later today to test this out more thoroughly! And once that is done. I believe we are good to go. 😄  

Thanks very much for your help and patience guys.

safaiyeh commented:
Appreciate all your hard work on this @YoliFD :)

nickmccomb commented:
Hey guys, is there an ETA on when this will be approved?

StagasaurusRex commented:
I was trying out the fork and had some syntax issues in RNCookieManagerIOS. I submitted a [PR](https://github.com/fanduel/cookies/pull/1) to fix these issues.

CHarnel commented:
Hey, appreciate all the hard work here, waiting for this to be approved & released - any ETA? :innocent:

StagasaurusRex commented:
I’ve been using this in production for a couple weeks now without issue. Seems ready to me.

safaiyeh commented:
I'll landed this! We can create issues for any other cases. A lot of good functionality.

safaiyeh commented:
@YoliFD We can take the left over work as separate issues, if you can resolve the conflicts we can merge this in :)

YoliFD commented:
Agreed @safaiyeh. I think we have addressed everything here except aligning the behaviour for wrong expiry date when trying to set cookies, which as you say we can tackle as a separate issue. 

I have just updated with master too and resolved the conflicts. Feel free to merge and release as you please and I will try to help with following up on resolved issues through this PR.

safaiyeh commented:
Thank you so much @YoliFD
